File: ios/CashuWallet/Core/WalletStore.swift
import Foundation

final class WalletStore {
    private let storage: StorageProtocol

    init(storage: StorageProtocol = UserDefaultsStorage()) {
        self.storage = storage
    }

    var activeMintURL: String? {
        get { value(forKey: StorageKeys.activeMintUrl) }
        set { setOptional(newValue, forKey: StorageKeys.activeMintUrl) }
    }

    func loadMints() -> [MintInfo] {
        let mints: [MintInfo] = value(forKey: StorageKeys.mints, legacyKeys: [StorageKeys.Legacy.mints]) ?? []
        let removed: [String] = value(forKey: StorageKeys.removedMintUrls) ?? []
        let removedIdentities = Set(removed.map(MintURLIdentity.normalized))
        return mints.filter { !removedIdentities.contains(MintURLIdentity.normalized($0.url)) }
    }

    func saveMints(_ mints: [MintInfo]) {
        set(mints, forKey: StorageKeys.mints)
    }

    /// CDK retains proofs after removal; they must not implicitly reconnect a mint.
    func isMintRemoved(url: String) -> Bool {
        let removed: [String] = value(forKey: StorageKeys.removedMintUrls) ?? []
        return removed.contains { MintURLIdentity.normalized($0) == MintURLIdentity.normalized(url) }
    }

    func setMintRemoved(url: String, removed: Bool) {
        let current: [String] = value(forKey: StorageKeys.removedMintUrls) ?? []
        let normalized = MintURLIdentity.normalized(url)
        var updated = current.filter { MintURLIdentity.normalized($0) != normalized }
        if removed { updated.append(normalized) }
        if updated != current { set(updated, forKey: StorageKeys.removedMintUrls) }
    }

    func loadBalancesByUnit() -> [String: UInt64] {
        value(forKey: StorageKeys.balancesByUnit) ?? [:]
    }

    func saveBalancesByUnit(_ balances: [String: UInt64]) {
        set(balances, forKey: StorageKeys.balancesByUnit)
    }

    func loadPendingReceiveTokens() -> [PendingReceiveToken] {
        value(
            forKey: StorageKeys.pendingReceiveTokens,
            legacyKeys: [StorageKeys.Legacy.pendingReceiveTokens]
        ) ?? []
    }

    func savePendingReceiveTokens(_ tokens: [PendingReceiveToken]) {
        set(tokens, forKey: StorageKeys.pendingReceiveTokens)
    }

    func loadCashuRequests() -> [CashuRequest] {
        value(forKey: StorageKeys.cashuRequests) ?? []
    }

    func loadSavedTokens() -> [String: String] {
        value(forKey: StorageKeys.savedTokens, legacyKeys: [StorageKeys.Legacy.savedTokens]) ?? [:]
    }

    func saveSavedTokens(_ tokens: [String: String]) {
        set(tokens, forKey: StorageKeys.savedTokens)
    }

    func loadPaymentPreimages() -> [String: String] {
        value(
            forKey: StorageKeys.paymentPreimages,
            legacyKeys: [StorageKeys.Legacy.paymentPreimages]
        ) ?? [:]
    }

    func savePaymentPreimages(_ preimages: [String: String]) {
        set(preimages, forKey: StorageKeys.paymentPreimages)
    }

    func loadMintQuoteTimestamps() -> [String: TimeInterval] {
        value(
            forKey: StorageKeys.mintQuoteTimestamps,
            legacyKeys: [StorageKeys.Legacy.mintQuoteTimestamps]
        ) ?? [:]
    }

    func saveMintQuoteTimestamps(_ timestamps: [String: TimeInterval]) {
        set(timestamps, forKey: StorageKeys.mintQuoteTimestamps)
    }

    func loadMintQuoteSchedules() -> [String: MintQuoteScheduleRecord] {
        value(forKey: StorageKeys.mintQuoteSchedules) ?? [:]
    }

    func saveMintQuoteSchedules(_ schedules: [String: MintQuoteScheduleRecord]) {
        set(schedules, forKey: StorageKeys.mintQuoteSchedules)
    }

    /// Last successful online keyset refresh per mint. Startup uses this to
    /// avoid contacting every configured mint on every app launch.
    func loadMintKeysetRefreshTimestamps() -> [String: TimeInterval] {
        value(forKey: StorageKeys.mintKeysetRefreshTimestamps) ?? [:]
    }

    func saveMintKeysetRefreshTimestamps(_ timestamps: [String: TimeInterval]) {
        set(timestamps, forKey: StorageKeys.mintKeysetRefreshTimestamps)
    }

    func loadProcessedNPCQuotes() -> [String] {
        value(forKey: StorageKeys.processedNPCQuotes) ?? []
    }

    func saveProcessedNPCQuotes(_ quoteIds: [String]) {
        set(quoteIds, forKey: StorageKeys.processedNPCQuotes)
    }

    func removeAllWalletData() {
        remove(keys: StorageKeys.walletDataKeys + StorageKeys.walletDataLegacyKeys)
        remove(keys: storage.keys(withPrefix: StorageKeys.walletDataPrefix))
    }

    /// One-way cleanup for stores obsoleted by the CDK 0.18 transaction
    /// lifecycle upgrade (local pending/claimed send records, async-melt
    /// tracking, melt fee notes, the old transaction cache). Idempotent.
    func purgeRetiredKeys() {
        remove(keys: StorageKeys.Retired.all)
    }

    private func value<T: Codable>(forKey key: String, legacyKeys: [String] = []) -> T? {
        if let value: T = try? storage.get(forKey: key) {
            return value
        }

        for legacyKey in legacyKeys {
            if let value: T = try? storage.get(forKey: legacyKey) {
                set(value, forKey: key)
                return value
            }
        }

        return nil
    }

    private func set<T: Codable>(_ value: T, forKey key: String) {
        do {
            try storage.set(value, forKey: key)
        } catch {
            AppLogger.wallet.error("Failed to save \(key): \(error)")
        }
    }

    private func setOptional<T: Codable>(_ value: T?, forKey key: String) {
        do {
            if let value {
                try storage.set(value, forKey: key)
            } else {
                try storage.remove(forKey: key)
            }
        } catch {
            AppLogger.wallet.error("Failed to update \(key): \(error)")
        }
    }

    private func remove(keys: [String]) {
        for key in Set(keys) {
            do {
                try storage.remove(forKey: key)
            } catch {
                AppLogger.wallet.error("Failed to remove \(key): \(error)")
            }
        }
    }
}

File: android/app/src/main/java/com/cashu/me/Core/WalletStore.kt
package com.cashu.me.Core

import android.content.Context
import kotlinx.serialization.KSerializer
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.builtins.MapSerializer
import kotlinx.serialization.builtins.serializer
import kotlinx.serialization.json.Json
import com.cashu.me.Core.Protocols.StorageKeys
import com.cashu.me.Models.CashuRequest
import com.cashu.me.Models.MintInfo
import com.cashu.me.Models.MintQuoteScheduleRecord
import com.cashu.me.Models.PendingReceiveToken
import com.cashu.me.Models.WalletTransaction

class WalletStore(
    context: Context,
    storeName: String = "wallet_store",
) : CashuRequestPersistence {
    private val store = DataStorePreferenceStore(context.applicationContext, storeName)
    private val json = Json { ignoreUnknownKeys = true; encodeDefaults = true }

    var activeMintURL: String?
        get() = store.string(StorageKeys.walletActiveMintUrl)
        set(value) = store.putString(StorageKeys.walletActiveMintUrl, value)

    fun loadMints(): List<MintInfo> {
        val removed = loadList(StorageKeys.walletRemovedMintUrls, String.serializer())
        return loadList(StorageKeys.walletMints, MintInfo.serializer()).filterNot { mint ->
            removed.any { com.cashu.me.Core.CDK.mintRemovalUrlsMatch(it, mint.url) }
        }
    }
    fun saveMints(mints: List<MintInfo>) = saveList(StorageKeys.walletMints, MintInfo.serializer(), mints)

    /** CDK retains proofs after removal; they must not implicitly reconnect a mint. */
    fun isMintRemoved(url: String): Boolean =
        loadList(StorageKeys.walletRemovedMintUrls, String.serializer())
            .any { com.cashu.me.Core.CDK.mintRemovalUrlsMatch(it, url) }

    @Synchronized
    fun setMintRemoved(url: String, removed: Boolean) {
        val current = loadList(StorageKeys.walletRemovedMintUrls, String.serializer())
        val updated = current.filterNot { com.cashu.me.Core.CDK.mintRemovalUrlsMatch(it, url) } +
            if (removed) listOf(url) else emptyList()
        if (updated != current) saveList(StorageKeys.walletRemovedMintUrls, String.serializer(), updated)
    }

    fun loadBalancesByUnit(): Map<String, Long> =
        loadMap(StorageKeys.walletBalancesByUnit, Long.serializer())
    fun saveBalancesByUnit(balances: Map<String, Long>) =
        saveMap(StorageKeys.walletBalancesByUnit, Long.serializer(), balances)

    fun loadPendingReceiveTokens(): List<PendingReceiveToken> = loadList(StorageKeys.walletPendingReceiveTokens, PendingReceiveToken.serializer())
    fun savePendingReceiveTokens(tokens: List<PendingReceiveToken>) =
        saveList(StorageKeys.walletPendingReceiveTokens, PendingReceiveToken.serializer(), tokens)

    /** txId → encoded token (re-display/reclaim of sent tokens, receive receipts). */
    fun loadSavedTokens(): Map<String, String> =
        loadMap(StorageKeys.walletSavedTokens, String.serializer())
    fun saveSavedTokens(tokens: Map<String, String>) =
        saveMap(StorageKeys.walletSavedTokens, String.serializer(), tokens)

    fun loadTransactions(): List<WalletTransaction> = loadList(StorageKeys.walletTransactions, WalletTransaction.serializer())
    fun saveTransactions(transactions: List<WalletTransaction>) =
        saveList(StorageKeys.walletTransactions, WalletTransaction.serializer(), transactions)

    fun loadPaymentPreimages(): Map<String, String> =
        loadMap(StorageKeys.walletPaymentPreimages, String.serializer())
    fun savePaymentPreimages(preimages: Map<String, String>) =
        saveMap(StorageKeys.walletPaymentPreimages, String.serializer(), preimages)

    fun loadMintQuoteTimestamps(): Map<String, Long> =
        loadMap(StorageKeys.walletMintQuoteTimestamps, Long.serializer())
    fun saveMintQuoteTimestamps(timestamps: Map<String, Long>) =
        saveMap(StorageKeys.walletMintQuoteTimestamps, Long.serializer(), timestamps)

    fun loadMintQuoteSchedules(): Map<String, MintQuoteScheduleRecord> =
        loadMap(StorageKeys.walletMintQuoteSchedules, MintQuoteScheduleRecord.serializer())
    fun saveMintQuoteSchedules(schedules: Map<String, MintQuoteScheduleRecord>) =
        saveMap(StorageKeys.walletMintQuoteSchedules, MintQuoteScheduleRecord.serializer(), schedules)

    fun loadProcessedNPCQuotes(): List<String> = loadList(StorageKeys.walletProcessedNPCQuotes, String.serializer())
    fun saveProcessedNPCQuotes(quotes: List<String>) =
        saveList(StorageKeys.walletProcessedNPCQuotes, String.serializer(), quotes)

    fun loadProcessedCashuRequests(): List<String> =
        loadList(StorageKeys.walletProcessedCashuRequests, String.serializer())
    fun saveProcessedCashuRequests(requestIds: List<String>) =
        saveList(StorageKeys.walletProcessedCashuRequests, String.serializer(), requestIds)

    fun loadProcessedNip17GiftWraps(): List<String> =
        loadList(StorageKeys.walletProcessedNip17GiftWraps, String.serializer())
    fun saveProcessedNip17GiftWraps(eventIds: List<String>) =
        saveList(StorageKeys.walletProcessedNip17GiftWraps, String.serializer(), eventIds)

    override fun loadCashuRequests(): List<CashuRequest> =
        loadList(StorageKeys.cashuRequests, CashuRequest.serializer()).map { it.withLegacyPaymentFallback() }
    override fun saveCashuRequests(requests: List<CashuRequest>) =
        saveList(StorageKeys.cashuRequests, CashuRequest.serializer(), requests.map { it.withLegacyPaymentFallback() })

    override var currentCashuRequestId: String?
        get() = store.string(StorageKeys.cashuRequestsCurrentId)
        set(value) = store.putString(StorageKeys.cashuRequestsCurrentId, value)

    internal fun snapshotWalletScopedData(): PreferenceSnapshot {
        val prefixKeys = store.keys().filter {
            it.startsWith(StorageKeys.walletDataPrefix) || it.startsWith(StorageKeys.npcDataPrefix)
        }
        return store.snapshot(StorageKeys.walletBoundaryKeys + prefixKeys)
    }

    internal fun restoreWalletScopedData(snapshot: PreferenceSnapshot) {
        store.restore(snapshot)
    }

    fun removeAllWalletData() {
        store.removeKeys(StorageKeys.walletBoundaryKeys)
        store.removePrefix(listOf(StorageKeys.walletDataPrefix, StorageKeys.npcDataPrefix))
    }

    /**
     * One-way cleanup for stores obsoleted by the CDK 0.18 transaction
     * lifecycle upgrade (local pending/claimed send records, async-melt
     * tracking, melt fee notes). Idempotent.
     */
    fun purgeRetiredKeys() {
        store.removeKeys(StorageKeys.retiredWalletKeys)
    }

    private fun <T> loadList(key: String, serializer: KSerializer<T>): List<T> {
        val raw = store.string(key) ?: return emptyList()
        return runCatching { json.decodeFromString(ListSerializer(serializer), raw) }.getOrDefault(emptyList())
    }

    private fun <T> saveList(key: String, serializer: KSerializer<T>, values: List<T>) {
        store.putString(key, json.encodeToString(ListSerializer(serializer), values))
    }

    private fun <T> loadMap(key: String, serializer: KSerializer<T>): Map<String, T> {
        val raw = store.string(key) ?: return emptyMap()
        return runCatching { json.decodeFromString(MapSerializer(String.serializer(), serializer), raw) }
            .getOrDefault(emptyMap())
    }

    private fun <T> saveMap(key: String, serializer: KSerializer<T>, values: Map<String, T>) {
        store.putString(key, json.encodeToString(MapSerializer(String.serializer(), serializer), values))
    }
}
